File: backend/python_engine/app/config/dsa_tables.py
"""
Resolve logical table names from DEV_MODE in backend/.env.

DEV_MODE=dev      → demo tables (instrument_data_demo, snapshot_demo, …)
DEV_MODE=production → production tables (instrument_data, instrument_snapshot, …)
"""

from __future__ import annotations

import os

TABLES: dict[str, tuple[str, str]] = {
    "company_profile": ("company_profile", "company_profile_demo"),
    "instrument_data": ("instrument_data", "instrument_data_demo"),
    "instruments": ("instruments", "instrument_demo"),
    "instrument_periods": ("instrument_periods", "instrument_period_demo"),
    "instrument_snapshot": ("instrument_snapshot", "snapshot_demo"),
}


def dev_mode() -> str:
    return os.environ.get("DEV_MODE", "dev").strip().lower()


def use_demo_tables() -> bool:
    mode = dev_mode()
    if mode in ("production", "prod"):
        return False
    if mode in ("dev", "demo", "local"):
        return True
    legacy = os.environ.get("DASHBOARD_USE_DEMO", "").strip().lower()
    if legacy in ("1", "true", "yes", "on"):
        return True
    if legacy in ("0", "false", "no", "off"):
        return False
    return mode != "production"


def table(logical: str) -> str:
    pair = TABLES.get(logical)
    if not pair:
        raise KeyError(f"Unknown logical DSA table: {logical}")
    prod, demo = pair
    return demo if use_demo_tables() else prod


def all_resolved() -> dict[str, str]:
    return {logical: table(logical) for logical in TABLES}

File: backend/app/Support/DsaTables.php
<?php

namespace App\Support;

use InvalidArgumentException;

/**
 * Resolve logical table names from DEV_MODE (dev → *_demo, production → production).
 */
final class DsaTables
{
    public static function devMode(): string
    {
        return strtolower(trim((string) config('dsa.dev_mode', 'dev')));
    }

    public static function useDemo(): bool
    {
        return (bool) config('dsa.use_demo_tables', true);
    }

    public static function name(string $logical): string
    {
        $tables = config("dsa.tables.{$logical}");
        if (! is_array($tables) || ! isset($tables['production'], $tables['demo'])) {
            throw new InvalidArgumentException("Unknown logical DSA table: {$logical}");
        }

        return self::useDemo() ? (string) $tables['demo'] : (string) $tables['production'];
    }

    /** @return array<string, string> logical => physical */
    public static function allResolved(): array
    {
        $out = [];
        foreach (array_keys(config('dsa.tables', [])) as $logical) {
            $out[$logical] = self::name($logical);
        }

        return $out;
    }
}
